### PR TITLE
count lexd entries using tree-sitter instead of by compiling

### PR DESCRIPTION
**Pull request by mr-martian**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -31,6 +31,10 @@ slog-envlogger = "2.1.0"
 slog-term = "2.4.0"
 tempfile = "3.0.1"
 tokio = { version = "1.2.0", features = ["process", "rt-multi-thread"] }
+tree-sitter = "0.19.3"
 
 [dev-dependencies]
 diesel_cli = { version = "1.4.1", default-features = false, features = ["sqlite"] }
+
+[build-dependencies]
+cc="*"
```
Comment (sushain97) [inline Cargo.toml:40]: Could you add a trailing newline here?

File: src/stats/lexd.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+use std::collections::HashSet;
+
+use tree_sitter::{Language, Parser, TreeCursor};
+
+use rocket_contrib::{json, json::JsonValue};
+
+use slog::Logger;
```
Comment (sushain97) [inline src/stats/lexd.rs:7]:
```suggestion
use tree_sitter::{Language, Parser, TreeCursor};
use rocket_contrib::{json, json::JsonValue};
use slog::Logger;
```
(current convention in this repo is to group together all the third-party deps)

File: src/stats/lexd.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+use crate::{models::StatKind, stats::StatsError};
+
+extern "C" {
+    fn tree_sitter_lexd() -> Language;
+}
+
+pub fn get_stats(_logger: &Logger, body: &str) -> Result<Vec<(StatKind, JsonValue)>, StatsError> {
+    let mut parser = Parser::new();
+    let language = unsafe { tree_sitter_lexd() };
+    parser.set_language(language).unwrap();
```
Comment (sushain97) [inline src/stats/lexd.rs:18]: Let's avoid `unwrap` whenever possible, it panics. Instead, could you use `map_err(...)?` here to map the error into a `StatsError` (which is just a wrapper type, you can add to it).

File: src/stats/lexd.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    let tree = parser.parse(body, None).unwrap();
```
Comment (sushain97) [inline src/stats/lexd.rs:19]: Same here, can use `ok_or_else(|| ...)?`.

File: src/stats/lexd.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        let mut walker2: TreeCursor = child.walk();
+        if child.kind() == "pattern_block" {
+            if child.child(0).unwrap().kind() == "pattern_start" {
+                patterns.insert("".to_string());
+            }
+            for line in child.children(&mut walker2) {
+                if line.kind() == "pattern_line" {
+                    pat_entries += 1;
+                } else if line.kind() == "identifier" {
+                    patterns.insert(body[line.byte_range()].to_string());
```
Comment (sushain97) [inline src/stats/lexd.rs:37]: Hmm, do you really need to use owned strings here? You should be able to take references into the existing string (declare the `Hash<&str>`). That should avoid a _ton_ of allocs.

**Review by sushain97 — CHANGES_REQUESTED** [review]
This is awesome! I'm pretty sure we can replace some of the other manual parsing with this as well. Just some minor nits.

File: src/stats/lexd.rs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+use slog::Logger;
+
+use crate::{models::StatKind, stats::StatsError};
+
+extern "C" {
+    fn tree_sitter_lexd() -> Language;
+}
+
+pub fn get_stats(_logger: &Logger, body: &str) -> Result<Vec<(StatKind, JsonValue)>, StatsError> {
+    let mut parser = Parser::new();
```
Comment (sushain97) [inline src/stats/lexd.rs:16]: If you wanted to have even more fun, you could wrap this in an object pool: https://github.com/zslayton/lifeguard and `reset` it for new parses.

**Comment by sushain97** [comment]
Sorry for the merge conflict! On the bright side, tests should be more reliably passing on `master` now.

File: .github/workflows/check.yml [hunk 7]
```diff
@@ -35,6 +35,8 @@ jobs:
     steps:
       - name: Checkout repo
         uses: actions/checkout@v2
+        with:
```
Comment (sushain97) [inline .github/workflows/check.yml:38]: You missed one spot on L13.

Comment (sushain97) [reply .github/workflows/check.yml:38]: Also the publish workflow

**Review by sushain97 — DISMISSED** [review]
🎉

File: src/stats/lexd.rs (hunk 6 above)
Comment (mr-martian) [reply src/stats/lexd.rs:14]: In principle the parser object could be created at start time and just kept around (that's what I do in apertium-lint, though that's not set up to be kept running).

**Comment by sushain97** [comment]
I think the linter is slightly unhappy with you: https://github.com/apertium/apertium-stats-service/pull/77/checks?check_run_id=2103056696.

LGTM otherwise.

**Comment by mr-martian** [comment]
yeah, I just forgot to run it before pushing

**Comment by mr-martian** [comment]
and I'd be happy to update the rest of tree-sitter-apertium if you want this approach for twol, cg, and rtx (would work for lexc too, I just haven't written the parser yet).

**Comment by sushain97** [comment]
happy to accept PRs! especially happy to replace any bespoke "parsing" I've written in this repo

**Review by sushain97 — APPROVED** [review]
I have to go fix coverage but until then

**Pull request merged**